### PR TITLE
Use secure gRPC api's instead of HTTP api's, use discovery, clean up codebase

### PR DESCRIPTION
**Pull request by romeovs**
This PR:

- Replaces the HTTP API by the gRPC API for the handler
- Uses discovery to discover the correct handler based on app ID, instead of using black magic for finding out regions
- Uses less magic in the data client, making it robuster
- Rewrites the codebase to be type-checked with flow type and linted like the rest of the TTN codebase
- Uses JSDoc in tandem with `documentation.js` to document the codebase
- Matches the `go-app-sdk` api better
- Adds full-stack tests that run on Travis

This has some breaking API changes but migrating should be easy.

### CODE REVIEW

**Comment by romeovs** [comment]
Currently, there's something wrong with the gRPC client and it won't connect to `discovery.thethings.network:1900` for instance, failing with a:

```
9356:error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown
```

I think this is caused by the server only supporting `TLSv1.x` and the gRPC client wanting to connect using `SSLv3` but I can't reproduce it locally. Maybe @htdvisser can have a look?

**Comment by johanstokking** [comment]
Let's not merge it until that issue gets fixed.

I don't know if we should support SSLv3 though. This issue doesn't happen with other endpoints (e.g. Handler)?

**Comment by johanstokking** [comment]
@romeovs 

- is Hylke's cipher suite update a fix? Travis doesn't seem to swallow it yet
- do we need SSLv3 on Discovery to make this work?
- why is TLS with the Handler working but not with Discovery?

**Comment by htdvisser** [comment]
- I'm still making some changes in the public discovery server to make it support it
- We don't need SSLv3
- Because the Handler was using ECDSA certificates, while our Let'sEncrypt certificates are RSA

**Comment by johanstokking** [comment]
Please reopen when tests pass and conflicts are resolved

**Comment by romeovs** [comment]
The tests are fixed!

**Comment by johanstokking** [comment]
@romeovs still blocked?

**Comment by romeovs** [comment]
@johanstokking it should be okay. It's just waiting for a review :)

File: .make/license.make [hunk 1]
```diff
@@ -0,0 +1,48 @@
+
```
Comment (johanstokking) [inline .make/license.make]: Remove file and references

File: README.md [hunk 2]
```diff
@@ -14,17 +14,80 @@ npm install --save ttn
...
+
+## Build
+
+To build the repository and transpile to ES5, run:
+
+```
+make build
+```
+
+## TODO's
```
Comment (johanstokking) [inline README.md]: Remove

**Review by johanstokking — CHANGES_REQUESTED** [review]

**Review by johanstokking — APPROVED** [review]

**Pull request merged**